core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala

```diff
@@ -193,8 +206,18 @@ private[spark] object HighlyCompressedMapStatus {
     } else {
       0
     }
+    val hugeBlockSizes = ArrayBuffer[Tuple2[Int, Byte]]()
+    if (numNonEmptyBlocks > 0) {
+      uncompressedSizes.zipWithIndex.foreach {
+        case (size, reduceId) =>
+          if (size > 2 * avgSize) {
```

Comment:
This should be configurable in two respects.
* minimum size before we consider something a large block : if average is 10kb, and some blocks are > 20kb, spilling them to disk would be highly suboptimal. (Unless I missed that check somewhere else).
* The fraction '2' should also be configurable - some deployments might be ok with high memory usage (machines provisioned accordingly) while others might need it to be more aggressive and lower.